File: ddd-basic/shop/src/main/java/basic/shop/order/command/domain/value/OrderNo.java
package basic.shop.order.command.domain.value;

import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;

import javax.persistence.Column;
import javax.persistence.Embeddable;
import java.io.Serializable;
import java.util.Objects;

@Embeddable
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@AllArgsConstructor
public class OrderNo implements Serializable {
    @Column(name = "order_number")
    private String number;

    @Override
    public boolean equals(Object o) {
        if(o == this)
            return true;
        if(!(o instanceof OrderNo))
            return false;
        OrderNo orderNo = (OrderNo) o;
        return Objects.equals(number, orderNo.number);
    }

    @Override
    public int hashCode() {
        return Objects.hashCode(number);
    }

    public static OrderNo of(String number) {
        return new OrderNo(number);
    }
}

File: spring-mvc-with-kotlin/order-api/src/main/kotlin/basic/orderapi/order/command/domain/value/OrderNo.kt
package basic.orderapi.order.command.domain.value

import java.io.Serializable
import java.util.*
import java.util.concurrent.ThreadLocalRandom
import javax.persistence.Column
import javax.persistence.Embeddable

@Embeddable
class OrderNo (
    @Column(name = "order_number")
    var number: String
): Serializable {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as OrderNo

        if (number != other.number) return false

        return true
    }

    override fun hashCode(): Int {
        return number.hashCode()
    }

    companion object {
        fun generate(): OrderNo {
            var randomNo = ThreadLocalRandom.current().nextInt(900000) + 100000;
            var number = String.format("%tY%<tm%<td%<tH-%d", Date(), randomNo)
            return OrderNo(number)
        }
    }
}